File: 62_unique_paths/uniquePaths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        arr = [[0 for x in range(n + 1)] for y in range(m + 1)]
        print(arr)
        return self.helper(m, n, arr)

    def helper(self, m: int, n: int, arr: List[List[int]]) -> int:
        if n == 1 and m == 1:
            return 1
        if n < 1 or m < 1:
            return 0

        if arr[m][n] != 0:
            return arr[m][n]

        arr[m][n] = self.helper(m - 1, n, arr) + self.helper(m, n - 1, arr)
        return arr[m][n]

File: 62_unique_paths/uniquePaths.ts
function uniquePaths(m: number, n: number): number {
  return helper(
    m,
    n,
    Array.from(Array(m + 1), () => new Array(n + 1))
  );
}

function helper(m: number, n: number, arr: number[][]): number {
  if (m == 1 && n == 1) return 1;
  if (m < 1 || n < 1) return 0;

  if (arr[m][n]) return arr[m][n];

  arr[m][n] = helper(m - 1, n, arr) + helper(m, n - 1, arr);

  return arr[m][n];
}
